### PR TITLE
KOA-5574 Handle internal & alpha colour tokens

### PR DESCRIPTION
**Pull request by marianeum**
We now have some new internal tokens and colour tokens with alpha values. These needed to be generated correctly to avoid exposing them.

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] Component `README.md`
+ [x] Tests
+ [x] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)

### CODE REVIEW

File: Backpack/src/main/res/values-night/backpack.semantic.color.xml [hunk 1]
```diff
@@ -18,76 +18,90 @@
 -->
 <resources>
 
-	<color name="bpkCanvas">#000000</color>
+	<color name="bpkCanvas">#ff000000</color>
```
Comment (marianeum) [inline Backpack/src/main/res/values-night/backpack.semantic.color.xml:21]: we just have the alpha channel included now

File: Backpack/src/main/res/values-night/backpack.semantic.color.xml [hunk 2]
```diff
@@ -18,76 +18,90 @@
...
-	<color name="bpkBackgroundSecondary">#1d1b20</color>
+	<color name="bpkBackgroundSecondary">#ff1d1b20</color>
 
-	<color name="bpkBackgroundAlternative">#000000</color>
+	<color name="bpkBackgroundAlternative">#ff000000</color>
 
-	<color name="bpkPrimary">#6d9feb</color>
+	<color name="bpkPrimary">#ff6d9feb</color>
+
+	<color name="__privateChipOnDarkNormalBackground">#ff0b121d</color>
```
Comment (marianeum) [inline Backpack/src/main/res/values-night/backpack.semantic.color.xml:93]: these are the new private tokens. in the view world moving them to a separate file won't help much as nobody accesses them via a file

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 3]
```diff
@@ -46,58 +49,78 @@ object BpkColor {
...
-      }
+      parseColors(source, resolveReferences = true) { it.isSemanticColor() && !it.isPrivate }.toBpkColors()
+  }
 
-    @OptIn(ExperimentalStdlibApi::class)
-    private fun BpkColorModel.hasSemanticSuffix(): Boolean {
-      val name = name.lowercase()
-      return semanticSuffixes.any { name.endsWith("_$it") && !name.endsWith("_on_$it") }
-    }
+  object Internal : BpkParser<Map<String, Any>, Map<String, BpkColors>> {
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:61]: there's now a third type of colours for internal tokens where we want dynamic tokens, but don't care about theming support

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 4]
```diff
@@ -46,58 +49,78 @@ object BpkColor {
...
+
   }
 
   private fun BpkColorModel.isMarcomms(): Boolean = name.startsWith("MARCOMMS_")
+
+  private fun BpkColorModel.isSemanticColor(): Boolean =
+    name != defaultReference && !hasSemanticSuffix() && !isMarcomms()
+
+  @OptIn(ExperimentalStdlibApi::class)
+  private fun BpkColorModel.hasSemanticSuffix(): Boolean {
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:98]: these just moved from the semantic block as they're also used for `internal`

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 5]
```diff
@@ -110,24 +133,21 @@ private fun parseColors(
...
-  val list = data
+  return data
     .map {
       BpkColorModel(
         name = it.key.trimName(),
-        defaultValue = it.value.getValue("value").trimColor(),
-        darkValue = it.value["darkValue"]?.trimColor(),
+        category = it.value.getValue("category").removeSuffix("-colors"),
+        isPrivate = it.key.startsWith("PRIVATE_"),
+        defaultValue = it.value.getValue("value"),
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:142]: got rid of trimming the color here as we have to modify it later in the process anyways and we no longer want to remove the suffix

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt [hunk 6]
```diff
@@ -156,8 +176,7 @@ private fun deprecationProperty() : PropertySpec =
     .build()
 
 @OptIn(ExperimentalStdlibApi::class)
-private fun String.toHexColor() =
-  "0xFF${uppercase()}"
+private fun String.toHexColor() = uppercase().run { "0x${substring(7)}${substring(1, 7)}" }
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkColors.kt:179]: RGBA -> ARGB

File: package.json [hunk 7]
```diff
@@ -61,7 +61,6 @@
     "merge-stream": "^2.0.0",
     "prettier": "^2.3.2",
     "svg2vectordrawable": "^2.4.0",
-    "through2": "^4.0.2",
-    "tinycolor2": "^1.4.1"
```
Comment (marianeum) [inline package.json:65]: one more dependency gone 🎉

File: gulpfile.js [hunk 8]
```diff
@@ -198,17 +199,22 @@ gulp.task('template:semanticColor', () => {
...
+        colorObject.name = `__${_.camelCase(tokenName)}`;
+        } else {
+          colorObject.name = `bpk${pascalCase(tokenName)}`;
+        }
 
         if (isSemanticColor(colorObject)) {
           const light = {
             ...colorObject,
-            value: tinycolor(colorObject.value).toHexString(),
+            value: asArgb(colorObject.value),
```
Comment (marianeum) [inline gulpfile.js:212]: before we.... took a hex string and... turned it into a hex string...??

**Review by olliecurtis — APPROVED** [review]

**Pull request merged**